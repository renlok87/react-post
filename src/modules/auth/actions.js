export const RECIEVE_AUTH = 'auth/RECIEVE_AUTH';
export const LOGIN = 'auth/LOGIN';
export const LOGOUT = 'auth/LOGOUT';
