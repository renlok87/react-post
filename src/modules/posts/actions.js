import moment from 'moment';
import { v4 as uuid } from 'uuid';

const scope = '@@posts';

export const CREATE_POST = `${scope}/CREATE_POST`;

export const createPost = payload => ({
  type: CREATE_POST,
  payload: {
    id: uuid(),
    createdAt: moment(),
    userId: null,
    text: '',
    replyToId: null,
    ...payload,
  },
});
