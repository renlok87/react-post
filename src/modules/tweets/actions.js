export const CREATE_TWEET = 'tweets/CREATE_TWEET';
