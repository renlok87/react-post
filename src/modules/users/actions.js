export const CREATE_USER = 'users/CREATE_USER';
