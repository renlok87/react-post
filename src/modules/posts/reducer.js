import * as actions from './actions';
import { combineReducers } from 'redux';

export const initialState = {
  allIds: [],
  byId: {},
};

export function allIds(state = initialState.allIds, action) {
  switch (action.type) {
    case actions.CREATE_POST:
      return [...state, action.payload.id];
    default:
      return state;
  }
}

export function byId(state = initialState.byId, action) {
  switch (action.type) {
    case actions.CREATE_POST:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    default:
      return state;
  }
}

export default combineReducers({
  allIds,
  byId,
});

// Selectors
export const getPostById = (state, id) => state.byId[id];
export const getRepliesById = (state, id) =>
  getAllPosts(state).filter(post => post.replyToId === id);
export const getAllPosts = state =>
  state.allIds.map(id => getPostById(state, id));
