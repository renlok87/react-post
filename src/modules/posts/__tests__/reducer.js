import reducer, {
  initialState,
  allIds,
  byId,
  getPostById,
  getAllPosts,
  getRepliesById,
} from '../reducer';
import * as actions from '../actions';

jest.mock('uuid', () => ({
  v4: () => '<id>',
}));

jest.mock('moment', () => () => '<datetime>');

describe('posts reducer', () => {
  it('returns initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
    expect(allIds(undefined, {})).toEqual(initialState.allIds);
    expect(byId(undefined, {})).toEqual(initialState.byId);
  });

  it('create new post', () => {
    const payload = { text: 'Test post' };
    const state = reducer(initialState, actions.createPost(payload));
    expect(state).toMatchSnapshot();
  });
});

describe('posts selectors', () => {
  it('gets posts by id', () => {
    const user = { id: 123, username: 'user' };
    const state = {
      byId: {
        [user.id]: user,
      },
    };

    expect(getPostById(state, user.id)).toEqual(user);
    expect(getPostById(state, 321)).toBeUndefined();
  });

  it('gets all posts', () => {
    const tweet1 = { id: 1, text: 'post 1' };
    const tweet2 = { id: 2, text: 'post 2' };
    const state = {
      allIds: [tweet1.id, tweet2.id],
      byId: {
        [tweet1.id]: tweet1,
        [tweet2.id]: tweet2,
      },
    };

    expect(getAllPosts(state)).toEqual([tweet1, tweet2]);
  });

  it('gets replies by id', () => {
    const post = { id: 1, text: 'post 1' };
    const reply = { id: 2, text: 'post 2', replyToId: 1 };
    const state = {
      allIds: [post.id, reply.id],
      byId: {
        [post.id]: post,
        [reply.id]: reply,
      },
    };

    expect(getRepliesById(state, post.id)).toEqual([reply]);
  });
});
