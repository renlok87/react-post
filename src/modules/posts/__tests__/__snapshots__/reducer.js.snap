// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tweets reducer create new post 1`] = `
Object {
  "allIds": Array [
    "<id>",
  ],
  "byId": Object {
    "<id>": Object {
      "createdAt": "<datetime>",
      "id": "<id>",
      "replyToId": null,
      "text": "Test post",
      "userId": null,
    },
  },
}
`;
